xbmc/cores/VideoPlayer/DVDDemuxers/DemuxTimeline.cpp

```diff
@@ -0,0 +1,297 @@
+#include "DemuxTimeline.h"
+
+#include <algorithm>
+
+#include "DVDClock.h"
+#include "DVDDemuxPacket.h"
+#include "DVDFactoryDemuxer.h"
+#include "DVDDemuxFFmpeg.h"
+#include "DVDInputStreams/DVDInputStreamFile.h"
+#include "DVDInputStreams/DVDFactoryInputStream.h"
+#include "filesystem/File.h"
+#include "filesystem/Directory.h"
+#include "MatroskaParser.h"
+#include "utils/log.h"
+#include "utils/StringUtils.h"
+
+CDemuxTimeline::CDemuxTimeline() {}
+
+CDemuxTimeline::~CDemuxTimeline() {}
+
+bool CDemuxTimeline::SwitchToNextDemuxer()
+{
+  if (m_curChapter->index + 1 == m_chapters.size())
+    return false;
+  CLog::Log(LOGDEBUG, "TimelineDemuxer: Switch Demuxer");
+  m_curChapter = &m_chapters[m_curChapter->index + 1];
+  m_curChapter->demuxer->SeekTime(m_curChapter->startSrcTime, true);
+  return true;
+}
+
+void CDemuxTimeline::Reset()
+{
+  for (auto &demuxer : m_demuxer)
+    demuxer->Reset();
+  m_curChapter = m_chapterMap.begin()->second;
+  if (m_curChapter->startSrcTime != 0)
+    m_curChapter->demuxer->SeekTime(m_curChapter->startSrcTime);
+}
+
+void CDemuxTimeline::Abort()
+{
+  m_curChapter->demuxer->Abort();
+}
+
+void CDemuxTimeline::Flush()
+{
+  m_curChapter->demuxer->Flush();
+}
+
+#define GET_PTS(x) ((x)->dts!=DVD_NOPTS_VALUE ? (x)->dts : (x)->pts)
+
+DemuxPacket* CDemuxTimeline::Read()
+{
+  DemuxPacket *packet = nullptr;
+  double pts = std::numeric_limits<double>::infinity();
+  double dispPts;
+
+  packet = m_curChapter->demuxer->Read();
+  packet && (pts = GET_PTS(packet));
+  while (
+    !packet ||
+    pts + packet->duration < DVD_MSEC_TO_TIME(m_curChapter->startSrcTime) ||
+    pts >= DVD_MSEC_TO_TIME(m_curChapter->stopSrcTime())
+  )
+  {
+    if (!packet || pts >= DVD_MSEC_TO_TIME(m_curChapter->stopSrcTime()))
+      if (!SwitchToNextDemuxer())
+        return nullptr;
+    packet = m_curChapter->demuxer->Read();
+    packet && (pts = GET_PTS(packet));
+  }
```

Comment:
I have rewritten this function multiple times until this loop was the result. I don't know how i could simplify it even more since you need to cover some special cases.